Description: disable failing test
Author: Marc Deslauriers <marc.deslauriers@canonical.com>
Forwarded: no

--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -102,7 +102,7 @@ ctests = mini-record-2 simple gc set_pkc
 	 fallback-scsv pkcs8-key-decode urls dtls-rehandshake-cert \
 	 key-usage-rsa key-usage-ecdhe-rsa mini-session-verify-function auto-verify \
 	 record-timeouts mini-dtls-hello-verify-48 mini-x509-default-prio \
-	 mini-x509-dual global-init-override tlsext-decoding rsa-psk-cb \
+	 mini-x509-dual tlsext-decoding rsa-psk-cb \
 	 rehandshake-switch-cert rehandshake-switch-cert-allow rehandshake-switch-cert-client \
 	 rehandshake-switch-cert-client-allow handshake-versions dtls-handshake-versions \
 	 dtls-max-record tls-max-record alpn-server-prec ocsp-filename-memleak \
